File: src/js/ts/js/src/DuckerCommand.ts
class DuckerCommand extends AbstractEffectCommand {

    private triggerGroup: SoundGroup;
    private ratio: number;
    private currentRatio: number;
    private diffRatioForAttackTimeBySample: number;
    private diffRatioForReleaseTimeBySample: number;

    constructor(triggerGroup: SoundGroup, ratio: number, attackTimeByMS: number = 0, releaseTimeByMS: number = 0) {
        super();
        this.triggerGroup = triggerGroup;
        this.ratio = ratio;
        this.currentRatio = 1;
        this.diffRatioForAttackTimeBySample = (1.0 - ratio) / (44100 * attackTimeByMS / 1000);
        this.diffRatioForReleaseTimeBySample = (1.0 - ratio) / (44100 * releaseTimeByMS / 1000);
    }

    public execute(pcmBufferArray: number[][]) {
        const numOfChannels: number = pcmBufferArray.length;
        for (let channel = 0; channel < numOfChannels; channel++) {
            const output = pcmBufferArray[channel];
            const size = output.length;

            if (this.triggerGroup.getCurrentPolyphony() > 0) {
                for (let i = 0; i < size; i++) {

                    if (this.currentRatio > this.ratio) {
                        this.currentRatio -= this.diffRatioForAttackTimeBySample;
                        if (this.currentRatio < this.ratio) {
                            this.currentRatio = this.ratio;
                        }
                    }
                    output[i] *= this.currentRatio;
                }
            } else {
                for (let i = 0; i < size; i++) {

                    if (this.currentRatio < 1) {
                        this.currentRatio += this.diffRatioForReleaseTimeBySample;
                        if (this.currentRatio > 1) {
                            this.currentRatio = 1;
                        }
                    }
                    output[i] *= this.currentRatio;
                }
            }

        }
    }
}

File: src/c/DuckerCommand.cpp
/**
*  Harmonia
*
*  Copyright (c) 2018 Hiiragi
*
*  This software is released under the MIT License.
*  http://opensource.org/licenses/mit-license.php
*/

#include "DuckerCommand.h"
#include "Harmonia.h"

DuckerCommand::DuckerCommand(SoundGroup* triggerGroup, float ratio, float attackTimeByMS, float releaseTimeByMS)
{
	_triggerGroup = triggerGroup;
	_ratio = ratio;
	_currentRatio = 1;
	_diffRatioForAttackTimeBySample = (1.0 - ratio) / (44100 * attackTimeByMS / 1000);
	_diffRatioForReleaseTimeBySample = (1.0 - ratio) / (44100 * releaseTimeByMS / 1000);
}


DuckerCommand::~DuckerCommand()
{
}


void DuckerCommand::execute(int* combinedBuffer, int size)
{
	if (_triggerGroup->get_current_polyphony() > 0)
	{
		for (int i = 0; i < size; i++)
		{
			if (_currentRatio > _ratio)
			{
				_currentRatio -= _diffRatioForAttackTimeBySample;
				if (_currentRatio < _ratio)
				{
					_currentRatio = _ratio;
				}
			}
			combinedBuffer[i] = (int)(combinedBuffer[i] * _currentRatio);

		}
	}
	else
	{
		for (int i = 0; i < size; i++)
		{
			if (_currentRatio < 1)
			{
				_currentRatio += _diffRatioForReleaseTimeBySample;
				if (_currentRatio > 1)
				{
					_currentRatio = 1;
				}
			}

			combinedBuffer[i] = (int)(combinedBuffer[i] * _currentRatio);
		}
	}
}
